src/txmempool.cpp

```diff
@@ -1043,17 +1043,23 @@ void CTxMemPool::TrimToSize(size_t sizelimit, std::vector<COutPoint>* pvNoSpends
 
     unsigned nTxnRemoved = 0;
     CFeeRate maxFeeRateRemoved(0);
-    while (!mapTx.empty() && DynamicMemoryUsage() > sizelimit) {
+    while (!mapTx.empty()) {
         indexed_transaction_set::index<descendant_score>::type::iterator it = mapTx.get<descendant_score>().begin();
 
+        CFeeRate removed(it->GetModFeesWithDescendants(), it->GetSizeWithDescendants());
+        // Skim away everything paying effectively zero fees, regardless of mempool size.
+        // Above that feerate, just trim until memory is within limits.
+        if (removed >= m_min_relay_feerate && DynamicMemoryUsage() <= sizelimit) break;
+
         // We set the new mempool min fee to the feerate of the removed set, plus the
         // "minimum reasonable fee rate" (ie some value under which we consider txn
         // to have 0 fee). This way, we don't allow txn to enter mempool with feerate
         // equal to txn which were removed with no block in between.
-        CFeeRate removed(it->GetModFeesWithDescendants(), it->GetSizeWithDescendants());
-        removed += m_incremental_relay_feerate;
-        trackPackageRemoved(removed);
-        maxFeeRateRemoved = std::max(maxFeeRateRemoved, removed);
+        if (removed >= m_min_relay_feerate) {
```

Comment:
I think this would have been slightly clearer to me as:

```c++
if (removed >= m_min_relay_feerate) {
    if (DynamicMemoryUsage() <= sizelimit) {
        break; // normal exit from loop
    }

    // only track fee rate for txs with descendant feerate above m_min_relay_feerate
    removed += incr; trackPackageRemoved(removed); mFRR = max(mFRR, removed);
}
```

Reply:
If we're concerned about the 2500 tx thing, we could limit that here couldn't we? ie:

```c++
int low_fee_removed = 0;
while (!empty() && low_fee_removed < 100) {
    ...
    if (removed < m_min_relay_feerate && DynamicMemoryUsage() <= sizelimit) {
        ++low_fee_removed;
    }
    ...
}
```

Reply:
> If we're concerned about the 2500 tx thing, we could limit that here couldn't we?

I think the problem with limiting this way is we're just putting it off - we still need to finish flushing everything before we build a block template or ATMP another transaction, otherwise we might build a low-feerate block template or let a tx spend from something 0-fee. So we could stop here, then mark that the mempool `needs_flush=true` (or something) and flush fully before we do either of those 2 things?

Reply:
What's the problem with building a low-feerate block template? The tx already propogated, so should be roughly fine for compact block relay; and if there are higher feerate txs, we'll select them anyway.

If we ATMP a child of the tx, and it CPFPs it enough to get above min_relay_feerate, then that's also fine, we just get the same behaviour with "accept A+B, drop A, accept C(pays for B)" as we do with "accept A+B, accept C, drop A".

Reply:
Adding a 0-fee package to a block template can not raise profitability, only potentially lower it(?)

Pretty easy to imagine different ancestors being evicted at different times, causing compact block mismatches. We could retain `blockmintxfee` though, right? Then all you're left with is some mismatches across pools potentially causing weirdness with individual vs package evaluation(dedup or not), but this is likely short-lived anyways.

Reply:
> What's the problem with building a low-feerate block template? 

So I think the argument for banning literal 0-fee packages is there is no gain from mining it, but it costs resources to include in the block. So it's probably not controversial to remove everything =0 fee, regardless of how many there are (?).

Beyond that (>0 but <minrelay), I don't have an argument other than the fact that a miner might expect `-blockmintxfee` to guarantee that the feerate of their block won't be lower than that.

Reply:
If we want to support miners being able to guarantee the feerate of their block won't be lower than `x`, then we shouldn't be removing/deprecating `-blockmintxfee` ?

I think it makes sense to have "ancestor fee(rate) is at 0sat? we're done here" logic; miners can always modify the fee if the tx is desired anyway...